Move linting to ruff

akx opened the pull request:
This PR switches linting (flake8, isort, <del>bandit</del>, yesqa, pygrep-hooks) for Pillow to [the `ruff` linter](https://github.com/charliermarsh/ruff).

* Running a full `pre-commit` run is much faster (45.15 seconds of CPU time to 8.62 seconds on my machine)
* It's easier to enable more rules if need be since the configuration is all in one place
* 👉  The only rule that may not be yet supported by `ruff` that had been previously enabled is `rst-backticks` from `pygrep-hooks`.
  * That being said, this PR still enables _more_ lints than had been in use from `pygrep-hooks` since it enables the whole `PGH` family.
* <del>👉 Instead of touching expressions like `x, y, l, r, w, a, d, f = metrics[ix]`, I opted to adding specific `noqa` directives to quiesce Ruff's "ambiguous name" for `l` (which could be an `1` or an `I` 😁)</del> No longer an issue via #6971.

hugovk commented on Makefile:
```diff
@@ -114,7 +114,4 @@ lint:
-	python3 -c "import isort" > /dev/null 2>&1 || python3 -m pip install isort
-	python3 -m black --target-version py37 .
-	python3 -m isort .
+	ruff --fix .
```
Why remove Black?

Please also make sure Ruff is installed.

akx replied:
Oops, oversight there.

akx replied:
That said, should this actually be `pre-commit run --all-files black ruff`?

hugovk commented on .pre-commit-config.yaml:
```diff
@@ -1,4 +1,10 @@
 repos:
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
```
Let's move Ruff after Black: Black might fix some things that Flake8 would complain about.

cclauss replied:
https://beta.ruff.rs/docs/usage
> Note that Ruff's pre-commit hook should run before Black, isort, and other formatting tools.

hugovk commented on .pre-commit-config.yaml:
```diff
@@ -1,4 +1,10 @@
 repos:
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
+    rev: 'v0.0.252'
```
To match others?
```suggestion
    rev: v0.0.252
```

hugovk commented on .pre-commit-config.yaml:
```diff
@@ -8,42 +14,12 @@ repos:
-    rev: 1.7.4
-    hooks:
-    - id: bandit
-      args: [--severity-level=high]
-      files: ^src/
```
How does Ruff deal with limiting the Bandit severity and files?

akx replied:
It doesn't, at present. The current configuration enables all Bandit rules except the three ignored ones, so in that sense it's stricter than before.

If Bandit should be disabled altogether for tests, that could be done with per-file-ignores.

cclauss replied:
charliermarsh/ruff#3239

hugovk commented on src/PIL/IptcImagePlugin.py:
```diff
@@ -18,10 +18,9 @@
 
 from . import Image, ImageFile
-from ._binary import i8
+from ._binary import i8, o8
```
What happened here? Does Ruff not support isort's Black profile?

akx replied:
isort is a bit looser than Ruff with these; Ruff wants either everything or nothing to be grouped like this, and isort doesn't always care.

hugovk replied:
Testing, isort always sorts this like the original, with or without the Black profile.

Anyway, I don't mind too much, as long as it remains relatively stable.

On that topic, maybe we should wait until Ruff is at "v1" or "stable", which looks to be close:

https://github.com/charliermarsh/ruff/issues/1992

cclauss replied:
Ruff has Black profile as the ONLY profile possible.

hugovk commented on setup.cfg:
```diff
@@ -63,13 +63,6 @@ tests =
     pytest-timeout
 
-[flake8]
-extend-ignore = E203
```
Does Ruff not have E203?

akx replied:
It doesn't.

hugovk commented on setup.py:
```diff
@@ -22,7 +22,7 @@
     version_file = "src/PIL/_version.py"
     with open(version_file, encoding="utf-8") as f:
-        exec(compile(f.read(), version_file, "exec"))
+        exec(compile(f.read(), version_file, "exec"))  # noqa: S102
```
Please limit Bandit to the `src` directory.

hugovk commented on src/PIL/BdfFontFile.py:
```diff
@@ -64,7 +64,7 @@ def bdf_char(f):
     bitmap = b"".join(bitmap)
 
-    [x, y, l, d] = [int(p) for p in props["BBX"].split()]
+    [x, y, l, d] = [int(p) for p in props["BBX"].split()]  # noqa: E741
```
Please can you check what `l` represents? Is there a better name?

Yay295 replied:
> The word BBX followed by the width in x (BBw), height in y (BBh), and x and y displacement
(BBox, BBoy) of the lower left corner from the origin of the character.

https://www.x.org/docs/BDF/bdf.pdf

So `l` and `d` should be the horizontal and vertical displacement. `x` and `y` could be renamed to `width` and `height` as well.

hugovk commented on src/PIL/Image.py:
```diff
@@ -767,7 +767,7 @@ def tobytes(self, encoder_name="raw", *args):
         data = []
         while True:
-            l, s, d = e.encode(bufsize)
+            l, s, d = e.encode(bufsize)  # noqa: E741
```
Please can you check what `l` represents? Is there a better name?

Yay295 replied:
`l`, `s`, and `d` are the length of the encoded data, an error code, and the encoded data.

hugovk commented on src/PIL/ImageFile.py:
```diff
@@ -530,12 +530,12 @@ def _encode_tile(im, fp, tile, bufsize, fh, exc=None):
                     # compress to Python file-compatible object
                     while True:
-                        l, s, d = encoder.encode(bufsize)
+                        l, s, d = encoder.encode(bufsize)  # noqa: E741
```
Please can you check what `l` represents? Is there a better name?

Yay295 replied:
`l`, `s`, and `d` are the length of the encoded data, an error code, and the encoded data.

hugovk commented on src/PIL/PcfFontFile.py:
```diff
@@ -86,7 +86,7 @@ def __init__(self, fp, charset_encoding="iso8859-1"):
         for ch, ix in enumerate(encoding):
             if ix is not None:
-                x, y, l, r, w, a, d, f = metrics[ix]
+                x, y, l, r, w, a, d, f = metrics[ix]  # noqa: E741
```
Please can you check what `l` represents? Is there a better name?
(And below)

Yay295 replied:
https://github.com/python-pillow/Pillow/blob/583dd3b4245649ddd367c2dcbbf2bf9fb7835017/src/PIL/PcfFontFile.py#L168

hugovk commented on src/PIL/TiffImagePlugin.py:
```diff
@@ -1845,7 +1845,7 @@ def _save(im, fp, filename):
         while True:
             # undone, change to self.decodermaxblock:
-            l, s, d = e.encode(16 * 1024)
+            l, s, d = e.encode(16 * 1024)  # noqa: E741
```
Please can you check what `l` represents? Is there a better name?

Yay295 replied:
`l`, `s`, and `d` are the length of the encoded data, an error code, and the encoded data.

hugovk reviewed (commented):
> * Running a full `pre-commit` run is much faster (45.15 seconds of CPU time to 8.62 seconds on my machine)

That's a good speedup! I get 5s -> 2s for `pre-commit run --all-files`, which is still welcome. On pre-commit.ci, it's [23.5s](https://results.pre-commit.ci/run/github/5171600/1677155148.Es8LdiESQVC7BxML-Cwxgg) -> [12.3s](https://results.pre-commit.ci/run/github/5171600/1677154839.tN83z3JTTTiN4YNogtkGjA).

akx commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+# Ignore Bandit in tests
+"Tests/*" = ["S"]
+"setup.py" = ["S"]
+# TODO: these files have some ambiguous `l` identifiers,
+#       which should be cleaned up.
+"src/PIL/BdfFontFile.py" = ["E741"]
+"src/PIL/Image.py" = ["E741"]
+"src/PIL/ImageFile.py" = ["E741"]
+"src/PIL/PcfFontFile.py" = ["E741"]
+"src/PIL/TiffImagePlugin.py" = ["E741"]
```
@hugovk Instead of diving deep into various font formats etc. right now, I opted to disable these codes for those files for now; I can gladly add subsequent PRs, but those would definitely be easier to review separately.

hugovk replied:
Sure. Maybe the per-line `noqa`s are better than per-file ignores, to prevent new ones being introduced?

akx replied:
@hugovk Well, I was planning on cleaning these up in subsequent PRs in the following days – and should new ones get introduced in that timeframe, it shouldn't be a big problem to clean them up then.

hugovk commented:
The addition of `pyproject.toml` has changed the build process from `setup.py`-based to `pyproject.toml`-based, leading to failures on MinGW:

https://github.com/python-pillow/Pillow/actions/runs/4253283135/jobs/7404119827

akx commented:
> The addition of `pyproject.toml` has changed the build process from `setup.py`-based to `pyproject.toml`-based, leading to failures on MinGW

Ah, should've known there are still uncivilized OSes like that... 😁 

Switched to `ruff.toml` instead 👍

akx commented:
@hugovk CI is green now, and #6971 will fix up the local variable names.

radarhere commented on Makefile:
```diff
@@ -115,6 +115,6 @@ lint:
-	python3 -c "import isort" > /dev/null 2>&1 || python3 -m pip install isort
 	python3 -m black --target-version py37 .
-	python3 -m isort .
+	which -s ruff || python3 -m pip install ruff
+	ruff --fix .
```
```suggestion
	python3 -c "import ruff" > /dev/null 2>&1 || python3 -m pip install ruff
	python3 -m ruff --fix .
```
On my machine, when I run `python3 -m pip install ruff`, it isn't automatically available to be run as `ruff`.

This is more flexible.

hugovk commented:
> Switched to `ruff.toml` instead 👍

Makes sense, we'll probably have to move to `pyproject.toml` at some point, but I agree it's out of scope for this. Please could you also add `check-toml` to `.pre-commit-config.yaml`?

aclark4life commented:
Yeah lots not switch, just add ruff … if possible/practical to do so.

cclauss commented:
If you do `ruff --format=github` then you get GitHub Annotations like:
![image](https://user-images.githubusercontent.com/3709715/225847266-19656f8a-ceb1-400d-8a83-b561b3402583.png)
Which some contributors might find quite helpful.  The problem is that when you run in pre-commit, you are unsure if `$CI` is true or false and you probably do not want `--format=github` unless it is true.

I run a separate GitHub that is superfast (<10 seconds) to quicky add annotations to code changes.
* internetarchive/openlibrary#7606

hugovk commented on ruff.toml:
I still need to review the rest of the PR, but let's name this file `.ruff.toml`. Our other config files also generally use the dot-prefix.

https://beta.ruff.rs/docs/configuration/

akx replied:
Done deal.

hugovk commented on .ruff.toml:
```diff
@@ -0,0 +1,28 @@
+    "ISC", # flake8-implicit-str-concat
+    "PGH", # pygrep-hooks
+    "RUF100",  # unused noqa (yesqa)
+    "S", # flake8-bandit
```
```suggestion
```

hugovk commented on .ruff.toml:
```diff
@@ -0,0 +1,28 @@
+    "W", # pycodestyle warnings
+    "YTT", # flake8-2020
+]
+ignore = [
+    # TODO: these could be enabled later
+    "S101", # use of assert
+    "S110", # try-except-pass
+    "S112", # try-except-continue
+]
```
```suggestion
```

hugovk commented on .ruff.toml:
```diff
@@ -0,0 +1,28 @@
+
+[isort]
+known-first-party = ["PIL"]
+
+[per-file-ignores]
+# Ignore Bandit in tests
+"Tests/*" = ["S"]
+"setup.py" = ["S"]
```
```suggestion
```

akx commented:
> > * Running a full `pre-commit` run is much faster (45.15 seconds of CPU time to 8.62 seconds on my machine)

I was measuring CPU time on purpose – wallclock time isn't affected that much, but the old tooling isn't able to utilize all cores on my machine (macOS 12.6.3, 2,4 GHz 8-Core Intel Core i9, etc.).

hugovk commented:
> * 👉  The only rule that may not be yet supported by `ruff` that had been previously enabled is `rst-backticks` from `pygrep-hooks`.
>   
>   * That being said, this PR still enables _more_ lints than had been in use from `pygrep-hooks` since it enables the whole `PGH` family.


https://beta.ruff.rs/docs/rules/#pygrep-hooks-pgh has four of the 10 hooks from https://github.com/pre-commit/pygrep-hooks:

- PGH001	`eval`	No builtin eval() allowed	
- PGH002	`deprecated-log-warn`	warn is deprecated in favor of warning	
- PGH003	`blanket-type-ignore`	Use specific rule codes when ignoring type issues	
- PGH004	`blanket-noqa`	Use specific rule codes when using noqa

`rst-backticks` is useful, let's add it back via the original `pygrep-hooks`.

hugovk commented on Makefile:
```diff
@@ -50,7 +50,7 @@ help:
 	@echo "  install-coverage   make and install with C coverage"
 	@echo "  lint               run the lint checks"
-	@echo "  lint-fix           run Black and isort to (mostly) fix lint issues"
+	@echo "  lint-fix           run ruff to (mostly) fix lint issues"
```
```suggestion
	@echo "  lint-fix           run Ruff to (mostly) fix lint issues"
```

akx commented on src/PIL/ImageFilter.py:
```diff
@@ -205,7 +205,7 @@ class UnsharpMask(MultibandFilter):
 
     .. _digital unsharp masking: https://en.wikipedia.org/wiki/Unsharp_masking#Digital_unsharp_masking
 
-    """  # noqa: E501
```
Long lines would be fixed by Black anyway, as far as it's able to. This is a long enough line that it can't be fixed by it.

hugovk replied:
Sure, but Flake8 complains if this `noqa` is missing, as does Ruff 0.0.259, whereas Ruff 0.0.260 is fine without it. 

```console
❯ flake8 --version
6.0.0 (mccabe: 0.7.0, pycodestyle: 2.10.0, pyflakes: 3.0.1) CPython 3.11.2 on Darwin
❯ flake8 src/PIL/ImageFilter.py
src/PIL/ImageFilter.py:206:89: E501 line too long (102 > 88 characters)
```
```console
❯ ruff --version
ruff 0.0.259

❯ ruff src/PIL/ImageFilter.py
src/PIL/ImageFilter.py:206:89: E501 Line too long (102 > 88 characters)
Found 1 error.
```
```console
❯ ruff --version
ruff 0.0.260

❯ ruff src/PIL/ImageFilter.py
```

Why has Ruff changed?

hugovk replied:
This is why: https://github.com/charliermarsh/ruff/pull/3663

akx replied:
Yeah, I think that's a reasonable change in Ruff's behavior TBH :)

cclauss commented:
What is the blocker or todo on this pull request?  It has been open for 9 weeks.

hugovk commented:
Partly reviewer time, partly waiting for a "stable" release: https://github.com/charliermarsh/ruff/issues/1992

And hopefully a stable release will be soon given the funding: https://astral.sh/blog/announcing-astral-the-company-behind-ruff

We're not in a rush to merge this, the speedup is negligible here.

I do want this merged though, so thank you to both of you for the PR and answers to my questions :)

akx commented:
Rebased, added a commit to upgrade to Ruff 275, fixed another autofix. 

Ruff will probably still not be stable-stable for a while (despite funding) so it might be nice to get this merged so we could maybe look at enabling more lints afterwards? :)

hugovk commented:
Like ruff, we're not in a rush :)

hugovk commented on .pre-commit-config.yaml:
```diff
@@ -1,51 +1,41 @@
+    rev: v0.0.275
+    hooks:
+      - id: ruff
+        args:
+          - --fix
```
Let's be consistent with the other hooks:

```suggestion
        args: [--fix]
```

And as suggested by https://github.com/astral-sh/ruff-pre-commit, should we exit with an error when it fixes, so that fixed commits aren't committed in the pre-fix state? 

```suggestion
        args: [--fix, --exit-non-zero-on-fix]
```

hugovk commented on _custom_build/backend.py:
```diff
@@ -1,6 +1,6 @@
 import sys
 
-from setuptools.build_meta import *  # noqa: F401, F403
```
Why does Flake8 warn about F401 when Ruff doesn't?

akx replied:
I suppose Flake8 warning about both F401 and F403 for a star import is overkill anyway.

hugovk commented:
That is helpful, thanks!

@akx Would you like to resolve the conflicts and update the config?

We've since removed `setup.cfg` (https://github.com/python-pillow/Pillow/pull/7484) and added some new pre-commit hooks including pyupgrade and flake-logging (https://github.com/python-pillow/Pillow/pull/7421).

akx commented:
@hugovk Done! I also have another branch that would stack on top of this to switch from `black` to [the Ruff formatter](https://astral.sh/blog/the-ruff-formatter) but that's a different concern so not muddling these waters with that.

akx commented:
The UP031 fixes would involve rewriting the `"something %dx%d" % some_tuple` style expressions in another way case by case - I'd leave them for an imminent follow up PR?

akx commented:
@hugovk @radarhere Could this maybe be good to go now? 😁 I could follow up with a PR to remove the six UP031 `noqa`s next...

radarhere commented on Makefile:
```diff
@@ -118,6 +118,6 @@ lint:
 	python3 -m black --target-version py38 .
-	python3 -m isort .
+	python3 -c "import ruff" > /dev/null 2>&1 || python3 -m pip install ruff
+	python3 -m ruff --preview --fix .
```
If stability is a concern, then wouldn't it be better to not use `--preview`? I have to imagine that any features that were in preview back when this PR was created are no longer in preview.

akx replied:
Sans `--preview`:

```
warning: Selection `LOG` has no effect because the `--preview` flag was not included.
warning: Selection `LOG` has no effect because the `--preview` flag was not included.
warning: Selection `LOG` has no effect because the `--preview` flag was not included.
```

akx replied:
(LOG was added via #7421.)

hugovk replied:
I think we should go for stability and avoid `--preview`. Any idea when LOG will leave preview?

akx replied:
I don't really think there's a big huge stability issue here, since whoever ends up updating the Ruff version in this repo will probably look at what they're doing?

Also, the more happy users a `preview` rule has, the better its chances are to graduate, I think 😁

hugovk replied:
Let's remove `--preview`, I'd prefer to keep things stable and not spend time double checking potentially-unstable changes every month.

akx replied:
Removed `--preview`, and also the `"LOG"` family of lints, because it would otherwise warn about it on every `--preview`less invocation.

radarhere commented:
I don't think I have strong opinions on the style changes seen here, so once @hugovk is ok with this, I'm happy for it to be merged.

hugovk reviewed (commented):
Please also delete `.flake8` (recently added in https://github.com/python-pillow/Pillow/pull/7484).

akx commented:
@hugovk Done, rebased, etc. :)

hugovk commented:
Thanks, a followup for UP031 would be good. And I'd prefer to keep Black rather than the Ruff formatter.
